File: 0102. Binary Tree Level Order Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#法一: 長度法
class Solution(object):
    def levelOrder(self, root):
        if not root:
            return []

        queue = collections.deque([root])
        result = []

        while queue:
            level = []
            for _ in range(len(queue)):
                cur = queue.popleft()
                level.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
            result.append(level)
        return result


#法二: recursion
class Solution(object):
    def levelOrder(self, root):
        levels = []
        self.helper(root, 0, levels)
        return levels

    def helper(self, node, level, levels):
        if not node:
            return
        if len(levels) == level:
            levels.append([])
        levels[level].append(node.val)

        self.helper(node.left, level + 1, levels)
        self.helper(node.right, level + 1, levels)

File: Java/7. Binary Tree/Traversal (Tree Level Order)/0102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// BFS
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Deque<TreeNode> que = new LinkedList<>();

        if (root != null) {
            que.add(root);
        }

        while (!que.isEmpty()) {
            List<Integer> levelValue = new ArrayList<>();
            int levelSize = que.size(); // 保存當前層的節點數量

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = que.poll();
                levelValue.add(node.val);

                if (node.left != null) {
                    que.add(node.left);
                }
                if (node.right != null) {
                    que.add(node.right);
                }
            }
            result.add(levelValue);
        }
        return result;
    }
}

// recursive
 class Solution {

     public List<List<Integer>> resList = new ArrayList<List<Integer>>();

     public List<List<Integer>> levelOrder(TreeNode root) {
         buildTreeRecursive(root, 0, resList); // 傳入 resList 而不是 result
         return resList;
     }

     public void buildTreeRecursive(TreeNode node, Integer level, List<List<Integer>> result) {
         // Base Case
         if (node == null) {
             return;
         }

         if (level == result.size()) {
             result.add(new ArrayList<>());
         }

         List<Integer> levelList = result.get(level);
         levelList.add(node.val);

         buildTreeRecursive(node.left, level + 1, result);
         buildTreeRecursive(node.right, level + 1, result);
     }
 }


// 1-1 Deque by NeetCode
// Notice: add value to queue or levelValue (List<Int> in result List)

class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> result = new ArrayList<>();
        Deque<TreeNode> que = new LinkedList<>();
        if (root != null) {
            que.add(root); // 1st time: add first root node on 1st level
        }

        while (!que.isEmpty()) {
            List<Integer> levelValue = new ArrayList<>();
            int levelSize = que.size(); // 保存當前層的節點數量; 1st time: 1

            for (int i = 0; i < levelSize; i++) {
                TreeNode node = que.poll();
                levelValue.add(node.val);

                if (node.left != null) {
                    que.add(node.left);
                }
                if (node.right != null) {
                    que.add(node.right);
                }
            }
            result.add(levelValue);
        }
        return result;
    }
}